### PR TITLE
Allow youtube video url to be added to a promotional feature

### PR DESCRIPTION
**Pull request by davidgisbey**
## Description

We're adding the ability for users to add a link to a youtube video in place or an image for promotional features.  This will allow us to use an embedded youtube video in place of an image. There's quite a lot of context in the Trello card (https://trello.com/c/hEEcf7QA/1007-add-a-video-field-to-whitehall-promotional-features) which is def worth reading.

This PR does the following:

1. Adds the `Add youtube urls to promotional features` permission so we can feature flag until the frontend work is done on Collections
2. Adds radio buttons which allow the user to choose between uploading an image or providing a Youtube URL
3. Adds a validation to check that an image or Youtube URL is present on a PromotionalFeatureItem (previously they could save without an image, but it would blow up with a 500 when published to the Publishing API)
4. Validates that the Youtube URL provided is valid
5. Pushes the Youtube video ID downstream to the Publishing API (not the whole URL)
6. Adds a JS module which hides the image fields when the youtube radio is selected and hides the youtube input when the image radio is selected.

### With permission

image

### Without permission

image

## Related PRs 

Collections - https://github.com/alphagov/collections/pull/3116
Publishing API ContentSchema changes - https://github.com/alphagov/publishing-api/pull/2237

## Trello card 

https://trello.com/c/hEEcf7QA/1007-add-a-video-field-to-whitehall-promotional-features

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

### CODE REVIEW

**Review by beccapearce — APPROVED** [review]
LGTM :)

File: app/assets/javascripts/admin_legacy/views/promotional_features/_form.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    },
+
+    addRadioButtonEventListeners: function addRadioButtonEventListeners () {
+      var form = this.form
+      var imageRadioButton = form.querySelector('#promotional_feature_promotional_feature_items_attributes_0_image_or_youtube_video_url_image')
+      var youtubeRadioButton = form.querySelector('#promotional_feature_promotional_feature_items_attributes_0_image_or_youtube_video_url_youtube_video_url')
+      var imageFields = form.querySelector('.image-fields')
+      var youtubeFields = form.querySelector('.youtube-video-url-fields')
+
+      imageRadioButton.addEventListener('click', function () {
```
Comment (kevindew) [inline app/assets/javascripts/admin_legacy/views/promotional_features/_form.js:35]: I'm not 100% sure but I think monitoring click might not be as effective as monitoring change on radio buttons, the latter feels more idiomatic but I am rusty in this area.

You can also probably manage this with a single event e.g:
```
var radioElements = document.querySelector('<radio buttons'>)

radioElements.addEventListener('change', function (event) {
  if (event.value == 'youtube_video_url') {
    // show youtube hide images
  } else {
    // show images hide youtube

  }
})

```

File: app/controllers/admin/promotional_feature_items_controller.rb [hunk 2]
```diff
@@ -56,15 +57,27 @@ def redirect_to_feature(notice = nil)
...
       :double_width,
       :image_cache,
+      :youtube_video_url,
+      :image_or_youtube_video_url,
       links_attributes: %i[url text _destroy id],
     )
   end
+
+  def clean_image_or_youtube_video_url_param
+    case promotional_feature_item_params.delete(:image_or_youtube_video_url).to_s
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/controllers/admin/promotional_feature_items_controller.rb:75]: Oh rightI could've sworn it did when i was manually testing. Let me have a play. I can always delete the associated image post save if required.

[the remark this answers is not in the document]
Comment (kevindew) [reply app/controllers/admin/promotional_feature_items_controller.rb]: Sure, I was actually expecting that model.assign_attributes(image: nil) wouldn't actually reset the field since submitting the form would essentially have an empty image form input (since you don't set an image on most form submissions).

But does look like it does:

```
[6] pry(main)> pfi.image.present?
=> true
[7] pry(main)> pfi.assign_attributes(image: nil)
=> nil
[8] pry(main)> pfi.image.present?
=> false
```

I wasn't thinking of Asset Manager but that is indeed a great shout, you're right we'd leave an orphaned asset. Your code looks good, but I'm a bit worried about the use of `.youtube_video_url.present?` which makes me wonder if we also leave orphaned images already when someone updates the image. I'd imagine we'd want the check to be more if the image url has changed?

File: app/models/promotional_feature_item.rb [hunk 3]
```diff
@@ -6,8 +6,15 @@ class PromotionalFeatureItem < ApplicationRecord
   validates :summary, presence: true, length: { maximum: 500 }
   validates_with ImageValidator, method: :image, size: [960, 640], if: :image_changed?
   validates :title_url, uri: true, allow_blank: true
+  validate :image_or_youtube_url_is_present
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/models/promotional_feature_item.rb:9]: Yea have added in logic to check both aren't present 👍

File: app/models/promotional_feature_item.rb [hunk 4]
```diff
@@ -6,8 +6,15 @@ class PromotionalFeatureItem < ApplicationRecord
...
 
   accepts_nested_attributes_for :links, allow_destroy: true, reject_if: ->(attributes) { attributes["url"].blank? }
 
   mount_uploader :image, ImageUploader
+
+private
+
+  def image_or_youtube_url_is_present
+    errors.add(:base, "Upload an image or add a Youtube URL") if image.blank? && youtube_video_url.blank?
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/models/promotional_feature_item.rb:35]: I did think about this, but thought that work arounds would probably be a bit overkill since this a sparsely touched feature and it'll likely only be a few days before it goes fully live.

File: app/assets/javascripts/admin_legacy/views/promotional_features/_form.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+      if (!imageRadioButtonChecked) {
+        var imageFields = form.querySelector('.image-fields')
+        imageFields.style.display = 'none'
+      }
+
+      if (!youtubeRadioButtonChecked) {
+        var youtubeFields = form.querySelector('.youtube-video-url-fields')
+        youtubeFields.style.display = 'none'
+      }
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/assets/javascripts/admin_legacy/views/promotional_features/_form.js:25]: There's a situation where both can be unchecked and both need to be hidden but i'll def steal using checked as a var 👍 thanks!

File: app/assets/javascripts/admin_legacy/views/promotional_features/_form.js (hunk 1 above)
Comment (davidgisbey) [reply app/assets/javascripts/admin_legacy/views/promotional_features/_form.js:35]: as it returns i had to use `querySelectorAll` and iterate over the results. Still simpler though for sure 👍

File: app/views/admin/promotional_features/new.html.erb [hunk 6]
```diff
@@ -1,3 +1,7 @@
+<% if current_user.can_add_a_youtube_urls_to_promotional_features? %>
+  <%= initialise_script "GOVUK.adminPromotionalFeaturesForm", selector: ".promotional_feature_item" %>
+<% end %>
```
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/views/admin/promotional_features/new.html.erb:3]: I could remove these and just return false if it can't find a radio fieldset on the page?

Atm it would blow up looking for radios on the page.

File: app/controllers/admin/promotional_feature_items_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/controllers/admin/promotional_feature_items_controller.rb]: Yea i think it wouldn't remove the image if we didn't mess with the params to manually set it to nil.

Great shout on change. I totally missed it'd still leave old images on change. I'll have a play now but i think something like

`delete_asset if legacy_url != current_url` 

would work

[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/controllers/admin/promotional_feature_items_controller.rb]: I've pushed up some changes if you'd like a 👀

File: app/views/admin/promotional_features/new.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply app/views/admin/promotional_features/new.html.erb:3]: Yup I think having the JS just abort would be good.

File: app/models/promotional_feature_item.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply app/models/promotional_feature_item.rb]: Yup makes sense.

File: app/assets/javascripts/admin_legacy/views/promotional_feature_items/_form.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  var promotionalFeatureItemsForm = {
+    init: function init (options) {
+      this.form = document.querySelector(options.selector)
+      this.toggleImageAndYoutubeFieldsVisibility()
+      this.addRadioButtonEventListeners()
+    },
+
+    toggleImageAndYoutubeFieldsVisibility: function toggleImageAndYoutubeFieldsVisibility () {
+      var form = this.form
+      var checked = form.querySelector('fieldset input[type=radio]:checked')
```
Comment (kevindew) [inline app/assets/javascripts/admin_legacy/views/promotional_feature_items/_form.js:14]: Bit of a nitpick, but this seems a bit too permissive as it'd catch any future radios. Could we put a class on the fieldset?

File: app/controllers/admin/promotional_feature_items_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevindew) [reply app/controllers/admin/promotional_feature_items_controller.rb]: Yup, that looks good. Nice one 👍

File: app/models/user.rb [hunk 8]
```diff
@@ -91,6 +92,10 @@ def can_preview_next_release?
     has_permission?(Permissions::PREVIEW_NEXT_RELEASE)
   end
 
+  def can_add_a_youtube_urls_to_promotional_features?
```
[the remark this answers is not in the document]
Comment (ollietreend) [reply app/models/user.rb]: @davidgisbey Sorry to reopen an old thread, but it's got me wondering: do we actually need a new permission for this specific feature?

I may just be out of the loop. But I was originally working under the assumption that there were already rules around who can/cannot update Promotional Features (e.g. the user has to be a GDS Admin). However now I've said it, I actually can't see anything to suggest the [Promotional Features](https://github.com/alphagov/whitehall/blob/186d8c50eb7e25f5b25ae0728e7fe02e70533fe3/app/controllers/admin/promotional_features_controller.rb) or [Promotional Feature Items](https://github.com/alphagov/whitehall/blob/186d8c50eb7e25f5b25ae0728e7fe02e70533fe3/app/controllers/admin/promotional_feature_items_controller.rb) controllers do enforce specific permissions. 😕

For example, the Organisations controller enforces permissions here:

https://github.com/alphagov/whitehall/blob/186d8c50eb7e25f5b25ae0728e7fe02e70533fe3/app/controllers/admin/organisations_controller.rb#L85-L92

with rules being defined here:

https://github.com/alphagov/whitehall/blob/186d8c50eb7e25f5b25ae0728e7fe02e70533fe3/lib/whitehall/authority/rules/organisation_rules.rb#L3-L16

and the fact that we don't also apply these rules to the Promotional Features is making me wonder if we have a bit of a loophole in our permissions model here? 🤔

Assuming that the URL/path for the Promotional Features page can be easily guessed _(spoiler: it can be)_, I think it means any authenticated user can change these regardless of their permissions or the organisation they belong to.

**TL;DR:** I don't think we need a user permission specifically to allow embedding YouTube videos. However I do think we're missing permission checks on these controllers in general.

File: app/assets/javascripts/admin_legacy/views/promotional_features/_form.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    addRadioButtonEventListeners: function addRadioButtonEventListeners () {
+      var form = this.form
+      var imageRadioButton = form.querySelector('#promotional_feature_promotional_feature_items_attributes_0_image_or_youtube_video_url_image')
+      var youtubeRadioButton = form.querySelector('#promotional_feature_promotional_feature_items_attributes_0_image_or_youtube_video_url_youtube_video_url')
+      var imageFields = form.querySelector('.image-fields')
+      var youtubeFields = form.querySelector('.youtube-video-url-fields')
+
+      imageRadioButton.addEventListener('click', function () {
+        imageFields.style.display = 'block'
```
[the remark this answers is not in the document]
Comment (ollietreend) [reply app/assets/javascripts/admin_legacy/views/promotional_features/_form.js]: 🙋🏻‍♂️ just to chime in here – I think an even nicer approach would be to set the `hidden` attribute of the element, rather than just setting the CSS styles.

e.g.

```
imageFields.hidden = true
```

and

```
imageFields.hidden = false
```

I think the `hidden` attribute has more semantic meaning – and it also means you don't have to do obscure things like setting `style.display = ''` and confusing future developers who might not realise that `''` empty strings reset CSS rules.

[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/assets/javascripts/admin_legacy/views/promotional_features/_form.js]: thanks have sorted it

File: app/models/user.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidgisbey) [reply app/models/user.rb]: Yea we should do this. The permission is just to hide it for now so we can merge in. It'll be removed after.
